### PR TITLE
Forming word groups with most frequent ID

### PR DESCRIPTION
**Pull request by abenezer-gelaw**
This PR contains code that creates word groups from DBSCAN labels. the ID of each cluster is the most frequent element in the label in the case it is unique, or the word alphabetically the first in case of multiple most frequent words. Unittest cases to test for edge cases have also been added along with an associated json file of test cases.

### CODE REVIEW

File: consistency/tool/clustering/clustering.py [hunk 1]
```diff
@@ -43,16 +44,74 @@ def extract_indices_dice(x, y):
...
+
+            return most_freq_words[0]  
+
+        for k in temp_dict.keys():
+            word_group = wg.WordGroup()
+            if k == -1:
+                word_group.id = "NOISE_WORDS"
+                word_group.kind = "WordGroup"
+                map = dict(Counter(temp_dict[k]))
+                for k, v in map.items():
```
Comment (shrutiparabgoogle) [inline consistency/tool/clustering/clustering.py:93]: NIY: Can we rename these `k, v` variables to what they are: `word, frequency`, there is already a variable k in the outer loop, which is creating some confusion.

File: consistency/tool/clustering/clustering.py [hunk 2]
```diff
@@ -43,16 +44,74 @@ def extract_indices_dice(x, y):
...
+    
+            if word_label in temp_dict:
+                temp_dict[word_label].append(self.words[j])
+                        
+            else:
+                temp_dict[word_label] = [self.words[j]]
+
+        word_groups = []
+
+        def find_clusterID(similar_list):
```
Comment (Krpate) [inline consistency/tool/clustering/clustering.py:77]: Any reason to make this a nested function? I am not too familiar with coding in python but wanted to bring your attention to this -> go/pystyle#nested-classes-functions-decision.

File: consistency/tool/clustering/clustering.py [hunk 3]
```diff
@@ -43,16 +44,74 @@ def extract_indices_dice(x, y):
...
+                similar_words = temp_dict[k]
+                id = find_clusterID(similar_words)
+
+                if id is None:
+                    continue
+                word_group.id = id
+                word_group.kind = "WordGroup"
+
+                map = dict(Counter(temp_dict[k]))
+                for k, v in map.items():
```
Comment (shrutiparabgoogle) [inline consistency/tool/clustering/clustering.py:106]: Same here

File: consistency/tool/clustering/clustering_test.json [hunk 4]
```diff
@@ -44,6 +44,45 @@
...
+                                            "Abandon":1,
+                                            "Collaborative":1,
+                                             "Duplication":1,
+                                             "Environments":1,
+                                             "Stragified":1
+                                          }
+                         }]
+   }, 
+   "duplicate-words": { 
+      "words": ["a", "a", "b", "b", "c"],
```
Comment (shrutiparabgoogle) [inline consistency/tool/clustering/clustering_test.json:65]: Since we have a condition of `len(word)>2` for a word to be in the list of valid_words,  can we update the examples here, just to make sure we have test cases with valid words

File: consistency/tool/clustering/clustering_test.json [hunk 5]
```diff
@@ -44,6 +44,45 @@
...
+                         "word_frequency": {
+                                            "Abandon":1,
+                                            "Collaborative":1,
+                                             "Duplication":1,
+                                             "Environments":1,
+                                             "Stragified":1
+                                          }
+                         }]
+   }, 
+   "duplicate-words": {
```
Comment (Krpate) [inline consistency/tool/clustering/clustering_test.json:64]: Can we add a test case where within a cluster there is ONLY ONE most common word?

**Review by shrutiparabgoogle — APPROVED** [review]
LGTM with few renames.

File: consistency/tool/clustering/clustering.py [hunk 6]
```diff
@@ -43,16 +44,74 @@ def extract_indices_dice(x, y):
...
+                    word_group.word_frequency[k] = v
+
+                word_groups.append(word_group)
```
Comment (Krpate) [inline consistency/tool/clustering/clustering.py:109]: nit: Line 103-107 are common in both if and else block. It can be moved outside to make it more compact.

File: consistency/tool/clustering/clustering.py (hunk 2 above)
Comment (abenezer-gelaw) [reply consistency/tool/clustering/clustering.py:77]: I see the style guide says it is fine to nest functions that are only used inside of a very limited scope. They say they are fine  with caveats. I am going to make it stay that way for consistency purposes (we have a nested function inside the clustering function, too.) unless you have strong opinions about them.

File: consistency/tool/clustering/clustering.py [hunk 7]
```diff
@@ -43,16 +44,74 @@ def extract_indices_dice(x, y):
...
         #if all labels are -1, DBSCAN detected no possible clusters. 
-        if np.count_nonzero(db[1] == -1) == db[1].size:
+        if np.count_nonzero(labels == -1) == labels.size:
             warnings.warn("There were no clusters detected. All words are unique.")
     
-        return db[1]
+        return labels
+
+    def create_word_groups(self):
+
```
[the remark this answers is not in the document]
Comment (abenezer-gelaw) [reply consistency/tool/clustering/clustering.py:59]: Formatted Using black formatter!
